### PR TITLE
Android

### PR DESCRIPTION
**Pull request by kubaau**

### CODE REVIEW

File: DolbyIO/DolbyIO.Android.xml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    </insert>
+    <insertValue value="dirs '$S(AbsPluginDir)/sdk-release-android'"/>
+    <insertNewline/>
+    <insert>}}}}</insert>
+  </baseBuildGradleAdditions>
+
+  <buildGradleAdditions>
+    <insert>
+      dependencies {
+        implementation("org.jetbrains.kotlin:kotlin-stdlib:1.7.20")
```
Comment (kubaau) [inline DolbyIO/DolbyIO.Android.xml:35]: Why is this necessary?

File: DolbyIO/DolbyIO.Android.xml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    <loadLibrary name="dolbyio_comms_sdk_android_cppsdk" 
+      failmsg="XML! libdolbyio_comms_sdk_android_cppsdk.so library not loaded and required!" />
+  </soLoadLibrary>
+
+  <baseBuildGradleAdditions>
+    <insert>
+      allprojects {
+        repositories {
+          repositories {
+            flatDir {
```
Comment (kubaau) [inline DolbyIO/DolbyIO.Android.xml:25]: Can I specify a local AAR without using Maven and without using flatDir? flatDir generates warnings.

File: DolbyIO/DolbyIO.Android.xml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+<root xmlns:android="http://schemas.android.com/apk/res/android">
+  <init>
+    <log text="DolbyIO plugin init"/>
+  </init>
+
+  <gradleProperties>
+    <insert>
+org.gradle.jvmargs=-Xmx1536M --add-exports=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.lang.reflect=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-exports=jdk.unsupported/sun.misc=ALL-UNNAMED
+ANDROID_BUILD_TOOLS_GRADLE_VERSION=com.android.tools.build:gradle:7.3.1
+ANDROID_TOOLS_BUILD_GRADLE_VERSION=com.android.tools.build:gradle:7.3.1
```
Comment (kubaau) [inline DolbyIO/DolbyIO.Android.xml:11]: What do all of these mean? Why is it 7.3.1 here and 7.4.1 in Engine?

File: DolbyIO/DolbyIO.Android.xml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    </insert>
+  </gradleProperties>
+
+  <soLoadLibrary>
+    <loadLibrary name="dolbyio_comms_sdk_android_cppsdk" 
+      failmsg="XML! libdolbyio_comms_sdk_android_cppsdk.so library not loaded and required!" />
+  </soLoadLibrary>
```
Comment (kubaau) [inline DolbyIO/DolbyIO.Android.xml:18]: Is this necessary?

File: README.md [hunk 5]
```diff
@@ -15,6 +15,7 @@ The plugin is compatible with the following operating systems:
 - macOS 10.14 or later if you use UE4
 - macOS 12.5 or later if you use UE5
 - Ubuntu 20.04 or later
+- Android ??? or later
```
Comment (kubaau) [inline README.md:18]: What are the Android platform requirements?

File: docs/docs/tutorial/android.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+This tutorial explains how to set up the plugin to target Android devices.
+
+## Prerequisites
+- JDK 17 or later
+
+## Setup
+1. Follow the [installation](installation) instructions for your development platform.
+2. Make sure that the Unreal Editor is closed.
+3. Download a plugin [release](https://github.com/DolbyIO/comms-sdk-unreal/releases) for Android and unpack the plugin to the `Plugins` folder. The release version should match your development platform release. Say yes to any overwrite prompts.
+4. Edit `{UnrealEngineRoot}/Build/Android/Java/gradle/gradle/wrapper/gradle-wrapper.properties` and modify `distributionUrl` to `https\://services.gradle.org/distributions/gradle-7.4.1-all.zip`
```
Comment (kubaau) [inline docs/docs/tutorial/android.md:16]: Is it possible to get rid of this step? Can changing the engine's Gradle distribution URL make other plugins not work?

File: docs/docs/tutorial/android.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+5. Open your project in Unreal Editor.
+6. Set `Project Settings -> Platforms -> Android SDK -> Location of JAVA` to the directory containing the required JDK version.
+7. Tick the `Project Settings -> Platforms -> Android -> Advanced Build -> Disable libc++_shared dependency validation in all dependencies` checkbox.
```
Comment (kubaau) [inline docs/docs/tutorial/android.md:19]: Same as above, except here I strongly feel that the Android libs should be build with static libc++ just like the C++ SDK libs are for Ubuntu with clang10. I don't think this requirement should exist as it disables warnings for the project which may be pertinent.

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -28,113 +156,27 @@ jobs:
...
           tag: ${{ env.CPP_SDK_TAG }}
-          asset_name: "cppsdk-${{ env.CPP_SDK_TAG }}-${{ env.BUILD_PLATFORM }}-universal.zip"
-      - name: Remove symlinks from cpp sdk libs
+          asset_name: "cppsdk-${{ env.CPP_SDK_TAG }}-${{ env.BUILD_PLATFORM }}.zip"
+      - name: Unpack AAR
         env:
-          LIB_PATH_UNIVERSAL: "${{ github.workspace }}/${{ env.CPP_SDK_DESTINATION }}/sdk-release/lib"
+          AAR_PATH_BASE: "${{ github.workspace }}/${{ env.CPP_SDK_DESTINATION }}/comms-sdk-android-cppsdk-"
+          AAR_PATH: "${{ env.AAR_PATH_BASE }}${{ env.CPP_SDK_TAG }}.aar"
+          AAR_NO_PREFAB_PATH: "${{ env.AAR_PATH_BASE }}no-prefab-${{ env.CPP_SDK_TAG }}.aar"
```
Comment (kubaau) [inline .github/workflows/build.yml:164]: Should be noprefab instead of no-prefab.

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -5,21 +5,149 @@ on:
   workflow_dispatch:
 
 env:
-  CPP_SDK_TAG: "2.5.4"
+  CPP_SDK_TAG: "2.6.0"
```
[the remark this answers is not in the document]
Comment (codlab) [reply .github/workflows/build.yml:8]: I think we can engage for something late summer (with a grain of salt here)

File: DolbyIO/DolbyIO.Android.xml (hunk 3 above)
Comment (codlab) [reply DolbyIO/DolbyIO.Android.xml:11]: we can definitely change to 7.4.1, I was making some changes locally and probably made a typo in my unreal env :p so mistakenly made 7.3.1 in that configuration. Let's bump it 👍

File: DolbyIO/DolbyIO.Android.xml (hunk 4 above)
Comment (codlab) [reply DolbyIO/DolbyIO.Android.xml:18]: yup !

what happens is that "normally" the cppsdk is the shared object that is provided for each platforms. However Android provides some layer of configurations that are normally transparent when using the maven dependency:

- media's so
- cppsdk's so
- and the upper layer of the "loader"

but since the cppsdk's so is not depending onto the upper layer (and that's the over way around - for clarity as regular android project), we need to make sure that this platform specific layer is loaded alongside the project's configuration - and this from the documentation should be done here

(but I don't exclude that later versions of the cppsdk will hide a bit more any shenanigans related to the platform's env)

File: DolbyIO/DolbyIO.Android.xml (hunk 2 above)
Comment (codlab) [reply DolbyIO/DolbyIO.Android.xml:25]: for curent versions, flatDir is fine, I don't see any other way right now. Once the cppsdk starts shipping the dependency to sonatype, it will be mandatory to remove those blocks anyway (and for intermediate versions, install in local maven repository : ~/.m2/repository/io/dolby etc... so that we avoid using aars)

File: DolbyIO/DolbyIO.Android.xml [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        implementation('io.dolby:comms-sdk-android-cppsdk-noprefab:2.6.0')
+      }
+      android {
+        packagingOptions {
+          pickFirst 'assets/model.dnr'
+        }
+      }
+    </insert>
+  </buildGradleAdditions>
+-->
```
[the remark this answers is not in the document]
Comment (codlab) [reply DolbyIO/DolbyIO.Android.xml:64]: A cleaner version will be to only encourage maven's oriented loading of libs.

mavenLocal() is actually pointing to ~/.m2/repository (and will be used for beta/intermediate cppsdk lib version)
google() to maven.google.com which provides every androidx libs etc...
mavenCentral() will be providing the production ready version of the cppsdk

File: DolbyIO/Source/Private/Utils/DolbyIOConversions.cpp [hunk 11]
```diff
@@ -156,6 +156,9 @@ namespace DolbyIO
 
 	spatial_audio_style ToSdkSpatialAudioStyle(EDolbyIOSpatialAudioStyle SpatialAudioStyle)
 	{
+#if PLATFORM_ANDROID
+		return spatial_audio_style::disabled;
+#endif
```
[the remark this answers is not in the document]
Comment (codlab) [reply DolbyIO/Source/Private/Utils/DolbyIOConversions.cpp:161]: I don't know. The only thing I know is that when I was loading the game and was joining the conference, a crash was happening in the set spatial position at the dvc level. Which we don't use if I'm not mistaken in the android sdk and I can't really debug that without the symbols from dvc. But maybe it works or maybe something else is happening.

It'll depend which one of us is quicker on that regard :p

File: android.sh [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+mkdir -p ${RELEASE_PATH_ARM64}
+mkdir -p ${RELEASE_PATH_X64}
+cp -r ${AAR_NAME}/prefab/modules/sdk/* ${RELEASE_PATH}
+cp ${AAR_NAME}/jni/arm64-v8a/libdolbyio_comms_media.so ${RELEASE_PATH_ARM64}
+cp ${AAR_NAME}/jni/arm64-v8a/libdolbyio_comms_sdk_android_cppsdk.so ${RELEASE_PATH_ARM64}
+cp ${AAR_NAME}/jni/x86_64/libdolbyio_comms_media.so ${RELEASE_PATH_X64}
+cp ${AAR_NAME}/jni/x86_64/libdolbyio_comms_sdk_android_cppsdk.so ${RELEASE_PATH_X64}
+
+if [[ "$AAR_NO_PREFAB_PATH" ]]; then
+  cp -r ${AAR_NO_PREFAB_PATH} ${RELEASE_PATH}
```
[the remark this answers is not in the document]
Comment (codlab) [reply android.sh:25]: to go in depth into the difference, it needs to be acknowledge that the cppsdk aar for android is pretty much normally something that needs to be consumed by another android project which unreal is not unfortunately :s

and so the end result library we would be shipping for android developers is something that would expose Kotlin implementation.

internally said final library would be compiled using cmake for the jni layer which then brings the question of "where do I get a so/a and the headers to compile the jni ?" and that's where the android environment is great now that cmake is shipped and usable out of the box for android projects: 

we can have "prefab" modules which are actually a specific set of folders in the aar where we can ship headers, sources and objects for the project's cmake to be able to load those as a known package but it comes to a cost where the lib itself is with some unneeded information that won't be changed if integrated as is later.

So we end up, for non regular project, to use:
- the prefab aar to get the actual headers and objects
- the non prefab aar to get the (almost) "final" libraries

I don't exclude that in the future, those steps will be simplified but once again it's more about the fact that unreal engine is not a regular project while being a bit like one. It overcomplexifies such things where it should normally be, in the best of both worlds doing everything by exposing a cmake project which could be consumed directly from a regular "platform's project" (cmake, gradle etc...)

File: docs/docs/tutorial/android.md [hunk 13]
```diff
@@ -0,0 +1,19 @@
+---
+sidebar_position: 9
+sidebar_label: Targetting Android
+title: Targetting Android
+---
+
+This tutorial explains how to set up the plugin to target Android devices.
+
+## Prerequisites
+- JDK 17 or later
```
[the remark this answers is not in the document]
Comment (codlab) [reply docs/docs/tutorial/android.md:10]: regarding the java version, we can't, the library we ship is requiring versions of the gradle wrapper and gradle environment which are not compatible with the decade old requisite of unreal engine

File: docs/docs/tutorial/android.md (hunk 6 above)
Comment (codlab) [reply docs/docs/tutorial/android.md:16]: We can't get rid of it. The distribution url version is required due. I don't see plugins where this would be a problem. I tested few like the store etc... but again, remember that unreal is actually an unstable half baked solution and that even today, I'm quite impressed it didn't fell apart already regarding Android (haven't dug much into the actual reasons behind that)

File: docs/docs/tutorial/android.md (hunk 7 above)
Comment (codlab) [reply docs/docs/tutorial/android.md:19]: The only reason is dvc and even there adding statically the libc++ will probably introduce memory management issues when it comes to unreal loading it's own and that dealing with memory ownership will be colliding.

In our case, the dynamic version of libc++ is normally ok as it happens that it's the only thing they did right out of the box : the ndk version is quite new (!)

File: DolbyIO/Source/Private/Utils/DolbyIOConversions.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (kubaau) [reply DolbyIO/Source/Private/Utils/DolbyIOConversions.cpp:161]: It didn't work for me either, but I didn't look into the crash at all really.

[the remark this answers is not in the document]
Comment (kubaau) [reply DolbyIO/Source/Private/Utils/DolbyIOConversions.cpp:161]: One other question that comes to mind: is there an Android equivalent of the desktop_app/showcase_app for desktop platforms? As in, a sample app that uses the C++SDK? desktop_app is very useful on desktops because it helps determine whether an issue is in plugin, app or SDK.

File: README.md (hunk 5 above)
Comment (kubaau) [reply README.md:18]: Most of the platform requirements do indeed come from Unreal. The only thing that C++SDK restricts further is the requirement for Windows 10 on UE4 (UE4 supports Windows 7). I'm just wondering, looking at https://docs.unrealengine.com/5.2/en-US/android-development-requirements-for-unreal-engine/ and https://docs.unrealengine.com/4.27/en-US/SharingAndReleasing/Mobile/Android/DeviceCompatibility/ , will there be any device that can run UE but cannot run the C++SDK?

File: DolbyIO/Source/Private/Utils/DolbyIOConversions.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (codlab) [reply DolbyIO/Source/Private/Utils/DolbyIOConversions.cpp:161]: we have a project for the actual android sdk. The interesting part is that I don't remember we even use the dvc's set position even in the regular project. Worth investigating sometime later for sure

File: README.md (hunk 5 above)
Comment (codlab) [reply README.md:18]: it would probably the other way around, having devices running c++sdk but not UE. But we are going to bump up the minimum version of Android to 23

File: .github/workflows/build.yml (hunk 8 above)
Comment (codlab) [reply .github/workflows/build.yml:164]: yup, little typo here

File: DolbyIO/DolbyIO.Android.xml (hunk 3 above)
Comment (kubaau) [reply DolbyIO/DolbyIO.Android.xml:11]: Setting these to 7.4.1 results in: `Minimum supported Gradle version is 7.5. Current version is 7.4.1. If using the gradle wrapper, try editing the distributionUrl in Z:\gradle\wrapper\gradle-wrapper.properties to gradle-7.5-all.zip`

Comment (codlab) [reply DolbyIO/DolbyIO.Android.xml:11]: My bad, I was reading it on mobile and I thought that discussion was about the distribution url completely. No we need to use AGP 7.3.1

File: README.md (hunk 5 above)
Comment (kubaau) [reply README.md:18]: This will likely mean that Android will not support UE4 which is stuck on 21 and does not build with 23, at least out of the box. Not sure if there is a market for this plugin to work on UE4 though.

File: scripts/android.sh [hunk 14]
```diff
@@ -0,0 +1,32 @@
+#!/bin/sh
```
Comment (kubaau) [inline scripts/android.sh:1]: This file needs updates regarding SDK version, shared aar and getting aars from Maven.

File: DolbyIO/Source/Private/Utils/DolbyIOErrorHandler.cpp [hunk 15]
```diff
@@ -26,6 +26,9 @@ namespace DolbyIO
 
 	void FErrorHandler::operator()(std::exception_ptr&& ExcPtr) const
 	{
+#if PLATFORM_ANDROID // why tho?
```
Comment (kubaau) [inline DolbyIO/Source/Private/Utils/DolbyIOErrorHandler.cpp:29]: This is the final issue assuming we can require to link libc++ dynamically.

**Review by Krysed — APPROVED** [review]
